### PR TITLE
Remove legacy validation

### LINKED ISSUE #4407: validation: default experimental_graphql_validation_mode to "new"
This should switch all router users to use only Rust validation. This is last step before entirely removing `experimental_graphql_validation_mode` config.

### LINKED ISSUE #4409: validation: remove experimental_graphql_validation_mode config
Completely transition to Rust validation in the router. These changes to include:
- [ ] set `graphql_validation`  to false when using `QueryPlannerConfig`
- [ ] remove metrics and traces that compare js vs rust validation errors as js errors do not exist anymore
- [ ] simplify validation paths for executable documents:
    - [ ] `ParsedDocumentInner` to work directly with `Valid<ExecutableDocument>` after an operation is received as part of the response
    - [ ] `QueryAnalysisLayer` should run overall validation
    - [ ] transition parsed schema to also use `Valid<Schema>`
    - [ ] verify the extra reparsing in `bridge_query_planner.rs` for filtered_query (is it necessary? can we reuse the existing `Valid<ExecutableDocument>`

### PR DESCRIPTION
**Pull request by Geal**
Fix #4407 
Fix #4409
This removes GraphQL validation from the query planner, to use the Rust version instead. Validation has now moved to the query analysis layer, which means we can remove a lot of code that was there to accommodate parsing and validating the query but not rejecting it outright, because we needed to compare the validation result with the planner's.
This will greatly reduce the load on the planner, which will now only be used for planning queries, not validating.
This new validation process has been running in production for months concurrently with the JavaScript version, allowing us to detect and fix any discrepancies in the new implementation. We now have enough confidence in the new Rust-based validation to entirely switch off the less performant, JavaScript validation.

Remaining issues:

- [x] usage reporting: since invalid queries are rejected in query analysis, inside the router service, then usage reporting for invalid queries is no longer reported, because the telemetry plugins handles usage reporting at the supergraph response level
- [x] persisted queries: apparently persisted queries use the query analysis transformation of supergraph requests, now one test fails on an invalid query, I have not yet looked at why
- [x] `@defer` directive: a test is failing because the directive is used in the query but not found by the schema. @lrlna what is apollo-compiler expecting here? Is `@defer` considered builtin, or does it need to be declared in the schema?

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Comment by lrlna** [comment]
> @defer directive: a test is failing because the directive is used in the query but not found by the schema. @lrlna what is apollo-compiler expecting here? Is @defer considered builtin, or does it need to be declared in the schema?

@Geal we do not define it as a built-in in apollo-compiler. The supergraphs that support defer, will have a defer definition already defined. And it looked like a bunch of tests <sup>[[1]](https://github.com/apollographql/router/blob/2c8115d7b67253686871473c4c4fe93b53476ad1/apollo-router/src/plugins/authorization/scopes.rs#L1180)</sup> <sup>[[2]](https://github.com/apollographql/router/blob/2c8115d7b67253686871473c4c4fe93b53476ad1/apollo-router/src/plugins/authorization/scopes.rs#L1180)</sup> <sup>[[3]](https://github.com/apollographql/router/blob/2c8115d7b67253686871473c4c4fe93b53476ad1/apollo-router/src/plugins/authorization/policy.rs#L1168)</sup> in the router already had them defined as part of the schema. I assume this test also needs it defined.

**Comment by Geal** [comment]
alright, that's an easy fix then

File: .changesets/feat_geal_remove_legacy_validation.md
Comment (lrlna) [inline .changesets/feat_geal_remove_legacy_validation.md:4]:
```suggestion
GraphQL query validation was initially performed by the query planner in JavaScript, which caused some performance issues. Here, we are introducing a new Rust-based validation process using `apollo-compiler` from the `apollo-rs` project.

This new validation process has been running in production for months concurrently with the JavaScript version, allowing us to detect and fix any discrepancies in the new implementation. We now have enough confidence in the new Rust-based validation to entirely switch off the less performant, JavaScript validation.
```

File: apollo-router/src/error.rs
Comment (lrlna) [inline apollo-router/src/error.rs:575]: Should this use `GraphQLError` or `DiagnosticsList`? 🤔  cc @SimonSapin

**Review by lrlna — COMMENTED** [review]
yesssss! tysm @Geal

File: apollo-router/src/error.rs
Comment (Geal) [reply apollo-router/src/error.rs:575]: we need something that implements `Serialize` and `Deserialize` so I used `GraphQLError`

Comment (SimonSapin) [reply apollo-router/src/error.rs:575]: For that purpose either works. `DiagnosticList` can be converted to `Vec<GraphQLError>` either when constructing this struct or later when using it. Doing it later allows printing with nice CLI formatting in `dbg!`, but requires carrying a `SourceMap` around somehow.

File: apollo-router/src/query_planner/bridge_query_planner.rs
Comment (BrynCooke) [inline apollo-router/src/query_planner/bridge_query_planner.rs:532]: Unwrap here, is this safe?

File: apollo-router/src/services/layers/query_analysis.rs
Comment (BrynCooke) [inline apollo-router/src/services/layers/query_analysis.rs:133]:
```suggestion
                        context.insert(OPERATION_NAME, operation_name).expect("cannot insert operation name into context; this is a bug");
```

File: apollo-router/src/query_planner/bridge_query_planner.rs
Comment (Geal) [reply apollo-router/src/query_planner/bridge_query_planner.rs:532]: we assume that the operations we generate are valid, but just in case I'll raise an error here

Comment (Geal) [reply apollo-router/src/query_planner/bridge_query_planner.rs:532]: bf7669828

**Review by BrynCooke — APPROVED** [review]

**Review by lrlna — APPROVED** [review]

**Comment by lrlna** [comment]
As an FYI, I know @goto-bus-stop had a change they need to run in our dual mode for a couple of days before we strip this out completely.

**Comment by Geal** [comment]
alright, I'm holding off on merging it for now

**Comment by goto-bus-stop** [comment]
We still need to approve & release the apollo-compiler validation fix and get it out in a public router release (just a patch version). Simon would be the best to review it but he's out today, I expect we'll have to wait a week before we can remove legacy validation entirely

**Comment by Geal** [comment]
updated to the latest dev now that #4510 is merged

**Comment by lrlna** [comment]
@Geal let me take a look at the metrics real quick

**Comment by lrlna** [comment]
@geal we haven't had any hiccups in the metrics since `1.40.1` released on monday. @goto-bus-stop the harness was matching with both implementations, yea?

**Comment by goto-bus-stop** [comment]
I am not sure if we have metrics from the new version yet specifically from the operations that were affected before (it's at least not in the nightly analytics rollup data yet). If we do, then we are good to go

**Comment by goto-bus-stop** [comment]
The option should also be removed from documentation: https://github.com/apollographql/router/blob/23656eac8beba4458f3e994538cf297403edc73c/docs/source/configuration/overview.mdx#L801-L811

**Comment by Geal** [comment]
@goto-bus-stop nice catch, thanks [746af65](https://github.com/apollographql/router/pull/4551/commits/746af65f505a88169700f18eb4667c5f6884c687)

File: apollo-router/src/query_planner/tests.rs
Comment (SimonSapin) [inline apollo-router/src/query_planner/tests.rs:425]: Is this still relevant after https://github.com/apollographql/router/pull/4701 ?

File: apollo-router/src/query_planner/bridge_query_planner.rs
Comment (SimonSapin) [inline apollo-router/src/query_planner/bridge_query_planner.rs:456]: Validation is hopefully faster now but still, do we want to run it 2~3 times per query?

Comment (goto-bus-stop) [reply apollo-router/src/query_planner/bridge_query_planner.rs:456]: Note also that some transformations do not produce valid documents today (an example in https://github.com/apollographql/router/pull/4938)

File: apollo-router/src/query_planner/tests.rs
Comment (goto-bus-stop) [reply apollo-router/src/query_planner/tests.rs:425]: Yeah it should be possible to go back to using `parse_test` here as it uses rust api schema now

File: apollo-router/src/plugins/demand_control/basic_cost_calculator.rs
Comment (goto-bus-stop) [inline apollo-router/src/plugins/demand_control/basic_cost_calculator.rs:140]: this was just a small mistake, the operation name isn't a type name, the type is always the root type (eg. for any `query` it's always `Query`, or whatever the schema has specified)

File: apollo-router/src/plugins/telemetry/metrics/snapshots/apollo_router__plugins__telemetry__metrics__apollo__test__apollo_metrics_multiple_operations.snap
Comment (SimonSapin) [inline apollo-router/src/plugins/telemetry/metrics/snapshots/apollo_router__plugins__telemetry__metrics__apollo__test__apollo_metrics_multiple_operations.snap:9]: Do we want to replace this metric with something else?

File: apollo-router/src/plugins/progressive_override/tests.rs
Comment (goto-bus-stop) [inline apollo-router/src/plugins/progressive_override/tests.rs:136]:
```suggestion
    let schema = crate::spec::Schema::parse_test(include_str!("./testdata/supergraph.graphql"), Default::default()).unwrap();
```
We should use `parse_test` everywhere in tests to ensure that the queries are pointing to the API schema rather than the supergraph schema, as they do in a running router

File: apollo-router/src/plugins/progressive_override/tests.rs
Comment (goto-bus-stop) [inline apollo-router/src/plugins/progressive_override/tests.rs:205]:
```suggestion
    let schema = crate::spec::Schema::parse_test(include_str!("./testdata/supergraph.graphql"), &Default::default()).unwrap();
```

File: apollo-router/src/plugins/progressive_override/tests.rs
Comment (goto-bus-stop) [inline apollo-router/src/plugins/progressive_override/tests.rs:276]:
```suggestion
    let schema = crate::spec::Schema::parse(include_str!("./testdata/supergraph.graphql"), &Default::default()).unwrap();
```

File: apollo-router/src/query_planner/testdata/defer_clause.graphql
Comment (goto-bus-stop) [inline apollo-router/src/query_planner/testdata/defer_clause.graphql:34]:
```suggestion

```
defer should not be declared in the supergraph but added by API schema, if there's a test failure the test should use `Schema::parse_test`

**Review by SimonSapin — APPROVED** [review]
Looks good overall, but I still think we should use `.to_json()` over `.to_string()` for parse errors: https://github.com/apollographql/router/pull/4551/files#r1525124171

The latter formats errors for a CLI-like output with a monospace font

File: apollo-router/src/services/supergraph/tests.rs
Comment (goto-bus-stop) [inline apollo-router/src/services/supergraph/tests.rs:1060]:
```suggestion
    let schema = Schema::parse_test(&new_schema, &configuration).unwrap();
```
I think it doesn't technically make a difference for this test, but lets use `Schema::parse_test` for consistency? especially since `schema.api_schema()` can silently do the wrong thing at the moment and we should catch those cases in tests

**Comment by SimonSapin** [comment]
Marked as approved to mean I don’t need to approve again after this change, but I do think the parse error thing should be changed.

**Review by goto-bus-stop — CHANGES_REQUESTED** [review]
Awesome 🤩  everything LGTM except a few places in tests where we should use `Schema::parse_test`

File: apollo-router/src/query_planner/bridge_query_planner.rs
Comment (goto-bus-stop) [reply apollo-router/src/query_planner/bridge_query_planner.rs]: We should probably keep this for now so we don't pass invalid documents to the query planner. When #4938 is fixed we can consider only validating if debug assertions are on and use `assume_valid` in prod?

**Comment by yanns** [comment]
We are using a plugin defining a  `supergraph_service` . Before this change, this method was called, we could do some request validation in it. After this change, it seems that this method is not called when the GraphQL query is invalid.
Can you confirm that this PR could change this behavior? Is it intended?

**Comment by abernix** [comment]
@yanns What was the request validation that you were doing at the supergraph service?  Something GraphQL-specific?

**Comment by yanns** [comment]
> @yanns What was the request validation that you were doing at the supergraph service? Something GraphQL-specific?

There are 2 use-cases:
- (1) as we have a public GraphQL API, we need to translate some error code. So we need a plugin that translates the error codes emitted by the router to "our" error codes. If the router rejects invalid GraphQL queries without calling plugins, we cannot modify the response.
- (2) we are using feature toggles. One GraphQL query does not use exactly the same GraphQL schema as another GraphQL query. So we need to validate requests depending on the GraphQL schema that is valid for this particular request. For this, I don't think that this PR is an issue. We configure the router with a GraphQL schema that has all possible fields. After the router has accepted the query, we validate it again with another GraphQL that is a subset of the one used by the router.

So I guess our main issue is that our plugin cannot adapt the response in case of invalid GraphQL queries.

**Comment by sushant3524** [comment]
Helllo Team, 
Really appreciate the work you guys are doing.
I have a use-case where I don't want any basic validation on router level. For that I did some changes in apollo-compiler and router to ignore validation error on rust , and for the most part it is working.
But for some queries, fragment types in particular are still giving validation error.
After exploring I have found out that these validation checks are coming from js (legacy).
I even tried using `experimental_graphql_validation_mode: new` in my yaml, but still it persists and validation errors are coming from Legacy.

To reproduce
Query is:
```
query getCall1($cursor: String, $size: Int) {
  call1: getCall1(cursor: $cursor, size: $size) {
    results {
      id
      payload {
        ...Payload1
        __typename
      }
      sender
      __typename
    }
    __typename
  }
}

fragment Payload1 on Payload1 {
  field1
  type
  ...Type1
  ...Type2
  __typename
}

fragment Type1 on Type1 {
  field2
  __typename
}

fragment Type2 on Type2 {
    field3
  __typename
}

```
and say `superschema doesn't have Type2 info.`
The error I am getting is 
```
{
	"errors": [
		{
			"message": "Cannot find fragment name \"Type2\" (provided fragments are: [Type1, Type3])",
			"extensions": {
				"code": "GRAPHQL_VALIDATION_FAILED"
			}
		}
	]
}
```
Any help is appreciated, thank you.

**Comment by goto-bus-stop** [comment]
@sushant3524 Accepting invalid queries in the router may cause all kind of unsupported glitches in the router and in the query planner. The router does not do validation in JS but I think the query planner may validate its own *outputs* at the end of query planning. Subgraphs also usually validate queries and if they return an error the router will propagate it. Everything the router does relies on the query being valid so trying to pass through something invalid is asking for trouble.

**Comment by sushant3524** [comment]
Thanks @goto-bus-stop  for the quick revert. Have reviewed much of your code in the router repository and it is a pleasure to finally engage with you.
I understand and accept the risks and will very well work on it. I just want that router doesn't validate through JS and as this PR removes it, I am not sure how and why JS is validating.

Is it because of [#4964](https://github.com/apollographql/router/pull/4964) ?
image
Or is it because router uses JS for some other usecase like maybe query_planner and there in JS if the query looks fishy they validate it again?

**Comment by sushant3524** [comment]
after searching I have found out that this is coming in bridge_query_planner from `PlannerMode`.
Are we working on removing dependency from JS for it?

**Pull request merged**
